--- /dev/null
+++ b/net/minecraft/util/datafix/fixes/BukkitNameFix.java
@@ -1,0 +_,28 @@
+package net.minecraft.util.datafix.fixes;
+
+import com.mojang.datafixers.DSL;
+import com.mojang.datafixers.DataFix;
+import com.mojang.datafixers.TypeRewriteRule;
+import com.mojang.datafixers.schemas.Schema;
+
+/**
+ * Project: Magma
+ * <p>
+ * Author: Malcolm (M1lc0lm)
+ * Date: 07.01.2023 - 19:25
+ */
+public class BukkitNameFix extends DataFix {
+
+  public BukkitNameFix(Schema outputSchema, boolean changesType) {
+    super(outputSchema, changesType);
+  }
+
+  @Override
+  protected TypeRewriteRule makeRule() {
+    return this.fixTypeEverywhereTyped(
+        "Player CustomName",
+        this.getInputSchema().getType(References.f_16772_),
+        (typed) -> typed.update(
+            DSL.remainderFinder(), EntityCustomNameToComponentFix::m_15407_));
+  }
+}
