--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -88,13 +_,15 @@
    private final VibrationListener.VibrationListenerConfig f_238787_;
    private final DynamicGameEventListener<Allay.JukeboxListener> f_238563_;
    private final SimpleContainer f_218303_ = new SimpleContainer(1);
-   public @Nullable BlockPos f_238682_;
+   @Nullable
+   public BlockPos f_238682_;
    public long f_238791_;
    private float f_218304_;
    private float f_218305_;
    private float f_238687_;
    private float f_238541_;
    private float f_238552_;
+   public boolean forceDancing = false; // CraftBukkit
 
    public Allay(EntityType<? extends Allay> p_218310_, Level p_218311_) {
       super(p_218310_, p_218311_);
@@ -106,6 +_,12 @@
       this.f_238563_ = new DynamicGameEventListener<>(new Allay.JukeboxListener(positionsource, GameEvent.f_238690_.m_157827_()));
    }
 
+   // CraftBukkit start
+   public void setCanDuplicate(boolean canDuplicate) {
+      this.f_19804_.m_135381_(Allay.f_238802_, canDuplicate);
+   }
+   // CraftBukkit end
+
    protected Brain.Provider<Allay> m_5490_() {
       return Brain.m_21923_(f_218301_, f_218297_);
    }
@@ -329,7 +_,7 @@
 
    public boolean m_7243_(ItemStack p_218387_) {
       ItemStack itemstack = this.m_21120_(InteractionHand.MAIN_HAND);
-      return !itemstack.m_41619_() && this.f_19853_.m_46469_().m_46207_(GameRules.f_46132_) && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_);
+      return !itemstack.m_41619_() && this.f_218303_.m_19183_(p_218387_) && this.m_245613_(itemstack, p_218387_) && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_19853_, this);
    }
 
    private boolean m_245613_(ItemStack p_252278_, ItemStack p_250405_) {
